service/src/main/java/bio/terra/workspace/db/FolderDao.java

```diff
@@ -0,0 +1,254 @@
+    }
+    var params = new MapSqlParameterSource();
+    Optional.ofNullable(displayName).ifPresent(name -> params.addValue("display_name", name));
+    Optional.ofNullable(description).ifPresent(d -> params.addValue("description", d));
+    if (Boolean.TRUE.equals(updateParent)) {
+      params.addValue(
+          "parent_folder_id",
+          Optional.ofNullable(parentFolderId).map(UUID::toString).orElse(DEFAULT_ROOT));
+    }
+    StringBuilder sb = new StringBuilder("UPDATE folder SET ");
+
+    sb.append(DbUtils.setColumnsClause(params));
+
+    sb.append(" WHERE workspace_id = :workspace_id AND id = :folder_id");
+    params
+        .addValue("workspace_id", workspaceId.toString())
+        .addValue("folder_id", folderId.toString());
+    try {
+      int rowsAffected = jdbcTemplate.update(sb.toString(), params);
+      return rowsAffected > 0;
+    } catch (DuplicateKeyException e) {
+      if (e.getMessage() != null
+          && e.getMessage()
+              .contains(
+                  "duplicate key value violates unique constraint \"folder_display_name_parent_folder_id_workspace_id_key\"")) {
+        throw new DuplicateFolderDisplayNameException(
+            String.format(
+                "Fails to update due to duplicate display name at the same folder level"));
+      }
+      throw e;
+    }
+  }
+
+  private boolean canFormCyclicCycle(UUID workspaceId, UUID folder1, UUID folder2) {
+    if (folder1 == null) {
+      return false;
+    }
+    if (folder1.equals(folder2)) {
+      return true;
+    }
+    Folder firstFolder = getFolder(workspaceId, folder1);
+    return canFormCyclicCycle(workspaceId, firstFolder.parentFolderId(), folder2);
+  }
+
+  @ReadTransaction
+  public Optional<Folder> getFolderIfExists(UUID workspaceId, UUID folderId) {
+    try {
+      return Optional.of(getFolder(workspaceId, folderId));
+    } catch (FolderNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  @ReadTransaction
+  public Folder getFolder(UUID workspaceId, UUID folderId) {
+    String sql =
+        """
+         SELECT workspace_id, id, display_name, description, parent_folder_id
+         FROM folder
+         WHERE id = :id AND workspace_id = :workspace_id
+       """;
+    var params = new MapSqlParameterSource();
+    params.addValue("id", folderId.toString());
+    params.addValue("workspace_id", workspaceId.toString());
+    try {
+      return DataAccessUtils.requiredSingleResult(
+          jdbcTemplate.query(sql, params, FOLDER_ROW_MAPPER));
+    } catch (EmptyResultDataAccessException e) {
+      throw new FolderNotFoundException(
+          String.format("Cannot find folder %s in workspace %s", folderId, workspaceId));
+    }
+  }
+
+  /**
+   * Gets a list of folders.
+   *
+   * @param parentFolderId when null, get all the folders in a workspace, otherwise, get the direct
+   *     subfolders of a given parent folder.
```

Comment:
List folders in parent folder is partially implemented

Done: Dao

Not done: service, API, controller unit test

Can you either take it out, or fully implement?

Reply:
there is a listFolders endpoint. It's in the controller unit test too. Also in the folder service

sorry there are too many new files.